worker/pubsub/subscriber.go

```diff
@@ -0,0 +1,215 @@
+		config:  config,
+		servers: make(map[string]RemoteServer),
+	}
+	subscription, err := config.Hub.Subscribe(pubsub.MatchAll, sub.forwardMessage)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sub.subscription = subscription
+	logger.Debugf("subscribing to details topic")
+	subscription, err = config.Hub.Subscribe(apiserver.DetailsTopic, sub.apiServerChanges)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sub.updates = subscription
+
+	err = catacomb.Invoke(catacomb.Plan{
+		Site: &sub.catacomb,
+		Work: sub.waitForDeath,
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return sub, nil
+}
+
+func (s *subscriber) waitForDeath() error {
+	logger.Tracef("wait for catacomb dying before unsubscribe")
+	defer s.subscription.Unsubscribe()
+	defer s.updates.Unsubscribe()
+
+	<-s.catacomb.Dying()
+	logger.Tracef("dying now")
+	return s.catacomb.ErrDying()
+}
+
+func (s *subscriber) apiServerChanges(topic pubsub.Topic, details apiserver.Details, err error) {
+	logger.Tracef("apiServerChanges: %#v", details)
+	// Make sure we have workers for the defined details.
+	if err != nil {
+		// This should never happen.
+		logger.Errorf("subscriber callback error: %v", err)
+		return
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	apiServers := set.NewStrings()
+	for id, apiServer := range details.Servers {
+		target := names.NewMachineTag(id).String()
+		apiServers.Add(target)
+		if target == s.config.Origin {
+			// We don't need to forward messages to ourselves.
+			continue
+		}
+
+		server, found := s.servers[target]
+		if found {
+			logger.Debugf("update addresses for %s", target)
+			server.UpdateAddresses(apiServer.Addresses)
+		} else {
+			logger.Debugf("new forwarder for %s", target)
+			newInfo := *s.config.APIInfo
+			newInfo.Addrs = apiServer.Addresses
+			server, err := s.config.NewRemote(RemoteServerConfig{
+				Hub:        s.config.Hub,
+				Origin:     s.config.Origin,
+				Target:     target,
+				Clock:      s.config.Clock,
+				APIInfo:    &newInfo,
+				APIOpen:    s.config.APIOpen,
+				WriterOpen: openMessageWriter,
+			})
+			if err != nil {
+				logger.Errorf("unable to add new remote server for %q, %v", target, err)
+				continue
+			}
+			s.servers[target] = server
+			s.catacomb.Add(server)
+		}
+	}
+	for name, server := range s.servers {
+		if !apiServers.Contains(name) {
+			logger.Debugf("%s no longer listed as an apiserver", name)
+			server.Kill()
+			err := server.Wait()
+			if err != nil {
+				logger.Errorf("%v", err)
+			}
+			delete(s.servers, name)
+		}
+	}
+	logger.Debugf("update complete")
+}
+
+func (s *subscriber) forwardMessage(topic pubsub.Topic, data map[string]interface{}) {
+	if data["origin"] != s.config.Origin {
+		// Message does not originate from the place we care about.
+		// Nothing to do.
+		logger.Tracef("skipping message %q as origin not ours", topic)
+		return
+	}
+	if local, ok := data["local-only"]; ok {
+		if value, ok := local.(bool); ok && value {
```

Comment:
Actually, I found that you can just go:
```go
  if data["local-only"] == true {
```
Because the default value is nil, it all kinda just works.